Add parameters to control page size in Parquet writer

etseidl opened the pull request:
This PR fixes #9615 

Adds two more parameters to the Parquet writer options objects to control page sizes.  One sets a target page size in bytes (defaults to 512 KiB), and a second to set a maximum number of rows per page (defaults to 20000, see [this](https://blog.cloudera.com/speeding-up-select-queries-with-parquet-page-indexes/) for the rationale behind this choice).  

~~This also removes the validation logic (and unit tests) from the row_group_size setters, since the page size is no longer fixed.  Perhaps validation could be moved to the build() function.~~

I've tried to follow the naming convention from #9677

Still needs help on python bindings and unit tests.

<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please mark your pull request as Draft.
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove it from "Draft" and make it "Ready for Review".
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review

   If assistance is required to complete the functionality, for example when the
   C/C++ code of a feature is complete but Python bindings are still required,
   then add the label `help wanted` so that others can triage and assist.
   The additional changes then can be implemented on top of the same PR.
   If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on the target branch, force push, or rewrite history.
   Doing any of these causes the context of any comments made by reviewers to be lost.
   If conflicts occur against the target branch they should be resolved by
   merging the target branch into the branch used for making the pull request.

8. Pull requests that modify cpp source that are marked ready for review 
   will automatically be assigned two cudf-cpp-codeowners reviewers.
   Ensure at least two approvals from cudf-cpp-codeowners before merging.

Many thanks in advance for your cooperation!

-->

GPUtester commented:
Can one of the admins verify this patch?

raydouglass commented:
okay to test

bdice commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -552,24 +568,22 @@ class parquet_writer_options {
-      "The maximum row group size cannot be smaller than the page size, which is 512KB.");
-    _row_group_size_bytes = size_bytes;
-  }
+  void set_row_group_size_bytes(size_t size_bytes) { _row_group_size_bytes = size_bytes; }
```
Should this validate the maximum row group size against the maximum page size?

I see the comment below suggesting to test parameters on initialization, but that isn't safe if this can be altered after initialization. Perhaps the best approach is to add validation to this method, and make the "builder" interface always sets the page size before the row group size. Then the error can be raised during the builder call OR after initialization.
```suggestion
  void set_row_group_size_bytes(size_t size_bytes)
  {
    CUDF_EXPECTS(
      size_bytes >= _target_page_size_bytes,
      "The maximum row group size in bytes cannot be smaller than the maximum page size in bytes.");
    _row_group_size_bytes = size_bytes;
  }
```

etseidl replied:
yes, I was definitely wanting some help on this logic.  my only concern is with confusion when modifying both row group and page sizes; the page size would have to be changed first.

vuule replied:
To me it makes the most sense to validate in `build()`. Allows any order and does not delay the validation too much (e.g. like validating in the reader would).

bdice replied:
@vuule Are you concerned about users calling `set_target_page_size_bytes` after the object is built? Is that allowed by the builder interface?

vuule replied:
You're right, that won't always work. 
We do have an equivalent interface in ORC. There, this is solved by only having a hard coded minimum for these values instead of comparing the two in the setters. We ensure that the size of the "row group" (called stripe in ORC) is not smaller than the size of "page" (called row group in ORC) in the getters instead of setters! Setters allow _stripes_ to be set to smaller size, but then getters for the _row group_ size return the minimum of the two. Effectively, setting stripe size low also sets the row group size to the same value.

etseidl replied:
@bdice @vuule Should I take a stab at reproducing the ORC logic?

bdice replied:
@etseidl Yes, that approach seems fine.

bdice commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -552,24 +568,22 @@ class parquet_writer_options {
-      "The maximum row group size cannot be smaller than the page size, which is 5000 rows.");
-    _row_group_size_rows = size_rows;
-  }
+  void set_row_group_size_rows(size_type size_rows) { _row_group_size_rows = size_rows; }
```
Should this validate the maximum row group size against the maximum page size?
```suggestion
  void set_row_group_size_rows(size_type size_rows)
  {
    CUDF_EXPECTS(
      size_rows >= _target_page_size_rows,
      "The maximum row group size in rows cannot be smaller than the maximum page size in rows.");
    _row_group_size_rows = size_rows;
  }
```

bdice commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -552,24 +568,22 @@ class parquet_writer_options {
+  /**
+   * @brief Sets the maximum page size, in bytes.
+   */
+  void set_target_page_size_bytes(size_t pgsz_bytes) { _target_page_size_bytes = pgsz_bytes; }
```
These arguments can be named the same as `set_row_group_size_bytes` above. The different function name makes it clear that this is a page size.
```suggestion
  void set_target_page_size_bytes(size_t size_bytes) { _target_page_size_bytes = size_bytes; }
```

bdice commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -552,24 +568,22 @@ class parquet_writer_options {
+  /**
+   * @brief Sets the maximum page size, in rows.
+   */
+  void set_target_page_size_rows(size_type pgsz_rows) { _target_page_size_rows = pgsz_rows; }
```
```suggestion
  void set_target_page_size_rows(size_type size_rows) { _target_page_size_rows = size_rows; }
```

bdice commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -482,6 +488,16 @@ class parquet_writer_options {
+  /**
+   * @brief Returns target page size, in bytes.
+   */
+  auto get_target_page_size_bytes() const { return _target_page_size_bytes; }
```
In some places this is called the "maximum page size" and in other places the "target page size." If pages cannot exceed this limit, I think renaming this to say "maximum" instead of "target" is appropriate.

etseidl replied:
done

bdice reviewed (commented):
Hi @etseidl! Thanks for your pull request! I have a few comments / suggestions.

bdice commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -482,6 +488,16 @@ class parquet_writer_options {
   auto get_row_group_size_rows() const { return _row_group_size_rows; }
 
+  /**
+   * @brief Returns target page size, in bytes.
```
```suggestion
   * @brief Returns maximum page size, in bytes.
```

bdice commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -482,6 +488,16 @@ class parquet_writer_options {
+  auto get_max_page_size_bytes() const { return _max_page_size_bytes; }
+
+  /**
+   * @brief Returns target page size, in rows.
```
```suggestion
   * @brief Returns maximum page size, in rows.
```

bdice commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -844,6 +886,16 @@ class chunked_parquet_writer_options {
   auto get_row_group_size_rows() const { return _row_group_size_rows; }
 
+  /**
+   * @brief Returns target page size, in bytes.
```
```suggestion
   * @brief Returns maximum page size, in bytes.
```

bdice commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -844,6 +886,16 @@ class chunked_parquet_writer_options {
+  auto get_max_page_size_bytes() const { return _max_page_size_bytes; }
+
+  /**
+   * @brief Returns target page size, in rows.
```
```suggestion
   * @brief Returns maximum page size, in rows.
```

bdice commented on cpp/src/io/parquet/page_enc.cu:
```diff
@@ -337,8 +339,13 @@ __global__ void __launch_bounds__(128)
                                : (values_in_page * 3 >= ck_g.num_values) ? 384 * 1024
                                                                          : 512 * 1024;
+
+      // override max_page_size if target is smaller
+      if (max_page_size > max_page_size_bytes) max_page_size = max_page_size_bytes;
```
We may want a better name for `max_page_size`, because it isn't clear how this value's meaning is different from `max_page_size_bytes`.
```
// TODO (dm): this convoluted logic to limit page size needs refactoring
```
@devavret Do you have suggestions on what should be done here?

etseidl replied:
It seems to me that intent here is to prevent wildly lopsided page sizes within a column chunk.  As the target page size shrinks, this should be less necessary.  Maybe `threshold_page_size` would be a better name?  Would something like the following make sense?
```c++
auto threshold_page_size = max_page_size_bytes;
if (threshold_page_size > SOME_CUTOFF_VALUE && (values_in_page * 3 >= ck_g.num_values)) {
  threshold_page_size = (values_in_page * 2 >= ck_g.num_values) ? max_page_size_bytes << 1
                                         : (max_page_size_bytes<<2)*3;
```
`SOME_CUTOFF_VALUE` would be 512*1024 to match the current code, but could be adjusted downward if need be.

bdice replied:
I think that sounds reasonable but would like a second reviewer's opinion on this section. cc: @vuule @devavret

devavret replied:
I was thinking more on the lines of adding a condition to the if block below that says: cut off the page if the remaining size of chunk is less than max_page_size_bytes. But that would require calculating the per chunk byte size beforehand.

I think the current changes are fine for now. For naming confusion, you can rename `max_page_size` to `this_page_max_size`.

BTW I think this PR only allows the user to make the pages smaller than they were going to be. Is there any need to allow larger pages?

devavret replied:
> It seems to me that intent here is to prevent wildly lopsided page sizes within a column chunk.

I think the real reason may be to avoid pages with very small number of values. I think we can remove this logic altogether because the final page is going to have at least 5000 values.

etseidl replied:
Ok, so for now I'll just rename the variable, but leave the TODO in place.

etseidl commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -699,6 +713,30 @@ class parquet_writer_options_builder {
+   * @param val The page size to use.
+   * @return this for chaining.
+   */
+  parquet_writer_options_builder& max_page_size_bytes(size_t val)
```
Understood.  This is a soft limit, since the current implementation operates a fragment-at-a-time, allowing a single fragment to over-fill a page.  I believe this is consistent with other parquet writers, such as parquet-mr.

etseidl commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -699,6 +713,30 @@ class parquet_writer_options_builder {
+   * @param val The page size to use.
+   * @return this for chaining.
+   */
+  parquet_writer_options_builder& max_page_size_rows(size_type val)
```
As with row groups, this is on row count, not value count.  I originally wasn't thinking of adding the row limit, but changed my mind later to allow cudf to have something akin to the row count limit from parquet-mr.  Will modify the documentation.

Question: Should I change the default to 0 and not limit on row count if that's the case?

devavret reviewed (commented):
The page size calculations still doesn't sit well with me because the size for the last few pages in a chunk is limited based on the number of values left to encode but actually it wants to limit the byte size. These are only loosely coupled as the values may be all null.

If we wanted to avoid pages with very small number of values, we don't need to worry anyway because the final page in a chunk is going to be at least 5000 values (fragment size).

etseidl commented:
@bdice @vuule @revans2 @devavret Thank you all for the feedback!  I think I've addressed all your suggestions.  Could you all take another look please?

devavret reviewed (approved):
Still some reservations about argument validation but since it's similar to existing code, I think it's out of scope for this PR.

vuule commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -567,9 +585,25 @@ class parquet_writer_options {
+  /**
+   * @brief Sets the maximum uncompressed page size, in bytes.
+   */
+  void set_max_page_size_bytes(size_t size_bytes) { _max_page_size_bytes = size_bytes; }
```
I think this should have a similar logic to `set_max_page_size_rows` and error below a hardcoded threshold.

etseidl replied:
@vuule What would you suggest as a lower bound?  Parquet itself is silent on this, is there a practical minimum size dictated by the compressors?  I would need to be able to go at least as low as 64K.

vuule replied:
4KB maybe? I'm pretty much making this up based on your requirement and assumed 512 limit for rows.

etseidl replied:
4KB works for me.  Done.

vuule commented on cpp/include/cudf/io/parquet.hpp:
```diff
@@ -567,9 +585,25 @@ class parquet_writer_options {
+  void set_max_page_size_rows(size_type size_rows)
+  {
+    CUDF_EXPECTS(
+      size_rows >= 5000,
```
This should be lower, IIUC 5K is the original size. ORC uses 512 here, maybe Parquet can do the same.

etseidl replied:
I kept 5000 since that's the size of a fragment.  I don't think setting lower than 5000 will have any effect.

vuule commented on cpp/src/io/parquet/page_enc.cu:
```diff
@@ -334,11 +336,16 @@ __global__ void __launch_bounds__(128)
+                                                                              : 512 * 1024;
+
+      // override this_max_page_size if the requested size is smaller
+      if (this_max_page_size > max_page_size_bytes) this_max_page_size = max_page_size_bytes;
```
```suggestion
      this_max_page_size = min(this_max_page_size, max_page_size_bytes);
```

etseidl replied:
done

vuule reviewed (approved):
Looks good, thank you for the contribution!

etseidl commented:
Thanks again everyone.  Not sure if this is the right place to bring this up, but it seems something changed in the CI build environment.  I keep getting build failures because of the arrow library referencing missing symbols.  Seems to be happening to other PRs as well.

bdice commented:
> Thanks again everyone. Not sure if this is the right place to bring this up, but it seems something changed in the CI build environment. I keep getting build failures because of the arrow library referencing missing symbols. Seems to be happening to other PRs as well.

Yes, this will be fixed as soon as possible. PR #10275 will resolve this.

bdice reviewed (approved):
This looks great @etseidl! Thanks for your quick iterations, and thanks to @vuule @devavret for the design advice.

vuule commented:
rerun tests

vuule commented:
rerun tests

vuule commented:
@gpucibot merge
